BUG: GroupBy.apply() returns different results if a different GroupBy method is called first

smithto1 opened the pull request:
- [x] closes #34656 
- [x] closes #34271 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

_Behavioural Changes_
_.apply()_
Calls  to `self._set_group_selection` have been replaced with `with _group_selection_context(self):` for `_agg_general`, `_make_wrapper`, and `nth`. 

Previously these calls to `self._set_group_selection` created a bug in `GroupBy.apply` where calling another method before `.apply` would change the output of `.apply`. This bug is now fixed. 


_Tests_ 
One new test is added to check that the output of `.apply` is constant whether another method is called on the same grouper first. 

Two existing tests were actually dependent on the old buggy-behaviour (i.e. they called GroupBy.sum first and then expected that GroupBy.apply(sum) would exclude the index columns from the results). All of these tests have been amended in a manner that enforces the new consistent output format while preserving the existing test. 

Both of the copy-pastable examples in the linked bug-reports are fixed.

smithto1 committed:
GroupBy.apply() calls self._reset_group_selection at the start. Errant tests updated

smithto1 committed:
Merge remote-tracking branch 'upstream/master' into issue34656

smithto1 committed:
gb.apply() now resets group selection so it always returns grouping columns as columns. updated tests that relied on previous behaviour

smithto1 committed:
test uses .drop() instead of selection

smithto1 committed:
wrote new tests

smithto1 committed:
rewrote test

smithto1 committed:
whatsnew

smithto1 committed:
restore if-stat in test_transform

smithto1 committed:
amended test

smithto1 committed:
restored test

smithto1 committed:
restored test

smithto1 committed:
amended test

smithto1 committed:
cleanup

pep8speaks commented:
Hello @smithto1! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`pandas/tests/groupby/test_apply.py`](https://github.com/pandas-dev/pandas/blob/f2a32f405c4a52c28827c516f34139b728541c8e/pandas/tests/groupby/test_apply.py):

> [Line 1027:55](https://github.com/pandas-dev/pandas/blob/f2a32f405c4a52c28827c516f34139b728541c8e/pandas/tests/groupby/test_apply.py#L1027): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

smithto1 committed:
trailing comma

smithto1 committed:
fixed test_to_latex

smithto1 committed:
added test to ensure .describe() keeps non-nuisance groupin columns

smithto1 committed:
minimized changes to exsiting tests

smithto1 committed:
add .describe() to whatsnew

jreback reviewed (approved):
lgtm @TomAugspurger of u would have a look

TomAugspurger commented on pandas/tests/groupby/test_function.py:
```diff
@@ -286,14 +286,47 @@ def test_non_cython_api():
-    expected_col = pd.MultiIndex(
-        levels=[["B"], ["count", "mean", "std", "min", "25%", "50%", "75%", "max"]],
-        codes=[[0] * 8, list(range(8))],
+    expected_col = pd.MultiIndex.from_product(
```
Just to confirm: this is the result you would get on master today if you called `df.groupby("A").describe()`, rather than doing a `.mad()` on the grouper before the `describe`?

smithto1 replied:
No. Currently on master, `describe` will always drop the grouping columns, even if these are non-nuisance columns, and it does not matter if another method is called first. `describe` always drops the grouping columns. 

This behaviour is changed with this PR, so that `.describe` will now keep non-nuisance columns in the output. 

Below is sample output from running the same code on master and this PR branch. 
```
import numpy as np
from pandas import DataFrame

df = DataFrame(
    [[1, 2, "foo"], [1, np.nan, "bar"], [3, np.nan, "baz"]], columns=["A", "B", "C"]
)

print( df.groupby('A').describe() )

# Output from master
#       B
#   count mean std  min  25%  50%  75%  max
# A
# 1   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0
# 3   0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN

# Output from PR branch
#       A                                        B
#   count mean  std  min  25%  50%  75%  max count mean std  min  25%  50%  75%  max
# A
# 1   2.0  1.0  0.0  1.0  1.0  1.0  1.0  1.0   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0
# 3   1.0  3.0  NaN  3.0  3.0  3.0  3.0  3.0   0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN
```

`test_non_cython_api` and `test_to_latex_multiindex` are the only tests impacted by this change, but the default behaviour is changing. One of the new tests I've added explicitly enforces the new default behaviour.

TomAugspurger replied:
Hmm, ok... @jreback thoughts on whether to call this a bugfix or API change? We document this behavior at https://pandas.pydata.org/docs/user_guide/groupby.html#aggregation, though we don't explicitly say that the grouping column is included.

----

Ignoring API-breaking issues for a minute and just thinking about desired output, shouldn't whether `A` is included depend on `as_index`? I would expect that with `as_index=True` then `A` would be excluded, and that with `as_index=False` it would be included. Is my intuition incorrect?

smithto1 replied:
> We document this behavior at https://pandas.pydata.org/docs/user_guide/groupby.html#aggregation, though we don't explicitly say that the grouping column is included.

The example of `grouped.describe()` under this link (screenshot below) would not actually change. 

![image](https://user-images.githubusercontent.com/3813175/87822355-384e5380-c869-11ea-9c31-5f59937e0ef1.png)

This example, reimplemented below, is grouping on two `str` columns which are nuisance columns for `.describe` so they are dropped. On the new branch they will still be dropped. The new branch only breaks the API if you group on non-nuisance columns (i.e. you group on floats or ints, then call `.describe`).

```
df = pd.DataFrame({'A': ['foo', 'bar', 'foo', 'bar','foo', 'bar', 'foo', 'foo'],
                   'B': ['one', 'one', 'two', 'three','two', 'two', 'one', 'three'],
                   'C': np.random.randn(8),
                   'D': np.random.randn(8)})

df.groupby(['A', 'B'], as_index=False).describe()

# Output from new branch is unchanged.
#       C                                          ...         D
#   count      mean       std       min       25%  ...       min       25%       50%       75%       max
# 0   1.0 -0.148328       NaN -0.148328 -0.148328  ...  1.315313  1.315313  1.315313  1.315313  1.315313
# 1   1.0  0.427926       NaN  0.427926  0.427926  ...  0.547973  0.547973  0.547973  0.547973  0.547973
# 2   1.0 -0.919795       NaN -0.919795 -0.919795  ...  0.292464  0.292464  0.292464  0.292464  0.292464
# 3   2.0 -0.335763  0.121342 -0.421565 -0.378664  ...  0.480334  0.679835  0.879337  1.078839  1.278340
# 4   1.0  0.708791       NaN  0.708791  0.708791  ... -2.003574 -2.003574 -2.003574 -2.003574 -2.003574
# 5   2.0  0.554458  1.536360 -0.531913  0.011272  ... -1.134286 -0.958744 -0.783202 -0.607660 -0.432118

# [6 rows x 16 columns]
```

smithto1 replied:
> shouldn't whether `A` is included depend on `as_index`? I would expect that with `as_index=True` then `A` would be excluded, and that with `as_index=False` it would be included. Is my intuition incorrect?

Currently on master, when grouping on a non-nuisance column, `as_index` has two effects:
    `as_index=True`  --> the groups are on the index and the grouping columns are excluded from the calculation results. 
    'as_index=False` --> the groups are not on the index and the grouping column is added to the calculation results. 

On the new branch, non-nuisance grouping columns are **always** included in the calculation results of `.describe`, so `as_index` only has one effect:
    `as_index=True`  --> the groups are on the index
    `as_index=False` --> the groups are not on the index

```
import numpy as np
from pandas import DataFrame

df = DataFrame(
    [[1, 2, "foo"], [1, np.nan, "bar"], [3, np.nan, "baz"]], columns=["A", "B", "C"]
)

print( df.groupby('A', as_index=True).describe() )

# master
#       B
#   count mean std  min  25%  50%  75%  max
# A
# 1   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0
# 3   0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN

# new branch
#       A                                        B
#   count mean  std  min  25%  50%  75%  max count mean std  min  25%  50%  75%  max
# A
# 1   2.0  1.0  0.0  1.0  1.0  1.0  1.0  1.0   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0
# 3   1.0  3.0  NaN  3.0  3.0  3.0  3.0  3.0   0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN

print( df.groupby('A', as_index=False).describe() )

# master
#       A                                        B
#   count mean  std  min  25%  50%  75%  max count mean std  min  25%  50%  75%  max
# 0   2.0  1.0  0.0  1.0  1.0  1.0  1.0  1.0   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0
# 1   1.0  3.0  NaN  3.0  3.0  3.0  3.0  3.0   0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN

# new branch
#       A                                        B
#   count mean  std  min  25%  50%  75%  max count mean std  min  25%  50%  75%  max
# 0   2.0  1.0  0.0  1.0  1.0  1.0  1.0  1.0   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0
# 1   1.0  3.0  NaN  3.0  3.0  3.0  3.0  3.0   0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN
```

smithto1 replied:
I have updated the new test, `test_describe_results_includes_non_nuisance_columns`, so it is paramtrized over `as_index=True/False`. This will enforce that the only affect of `as_index` on `describe` is to change the index from the groups to a range index, with no other side effects.

smithto1 replied:
#28549 is related. 

I did look at having `.apply` always suppress the grouping column but this breaks vastly more tests.

smithto1 replied:
Ignore my previous comment (now deleted) if you read them. 

I've updated the PR so that the behaviour of `.describe()` is unchanged. 

```

import numpy as np
from pandas import DataFrame

df = DataFrame(
    [[1, 2, "foo"], [1, np.nan, "bar"], [3, np.nan, "baz"]], columns=["A", "B", "C"]
)

print( df.groupby('A', as_index=True).describe() )
print( df.groupby('A', as_index=False).describe() )

# Output from PR branch (same as master)
#       B
#   count mean std  min  25%  50%  75%  max
# A
# 1   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0
# 3   0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN
#       A                                        B
#   count mean  std  min  25%  50%  75%  max count mean std  min  25%  50%  75%  max
# 0   2.0  1.0  0.0  1.0  1.0  1.0  1.0  1.0   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0
# 1   1.0  3.0  NaN  3.0  3.0  3.0  3.0  3.0   0.0  NaN NaN  NaN  NaN  NaN  NaN  NaN
```

smithto1 replied:
Following this update the `.describe` method is no longer affected by this PR. 

https://github.com/pandas-dev/pandas/pull/35314#issuecomment-664193224

TomAugspurger reviewed (approved):
Looks good, if my understanding on that point is correct.

smithto1 committed:
parametrize test over as_index=T/F

smithto1 committed:
restored .describe to old behaviour

smithto1 committed:
Merge remote-tracking branch 'upstream/master' into issue34656

smithto1 committed:
Merge remote-tracking branch 'upstream/master' into issue34656-temp

smithto1 committed:
restoring test_function.py to master

smithto1 committed:
added comment

smithto1 committed:
Merge branch 'issue34656-temp' into issue34656

smithto1 commented:
@TomAugspurger I think all the issues are resolved with this PR. Can you do a second review?

jreback commented on pandas/core/groupby/groupby.py:
```diff
@@ -1623,11 +1625,17 @@ def ohlc(self) -> DataFrame:
+        if self.axis == 1:
+            return result.T
+        # GH 34656 self.apply() will return non-nuisance grouping columns, but
+        # we remove them from describe if as_index=True
```
why are we doing this compatiblity? this seems like a step backwards

smithto1 replied:
The current behaviour of `GroupBy.describe` is superior. It doesn't make sense to return a non-nuisance grouping column in the output of `GroupBy.describe()` because that column will, by construction, only have one value in each group and returning the *mean, std, min, 25%, 50%, 75%, max* of a single value is basically meaningless. 

If the user calls `GroupBy.describe` with `as_index=False` they will get a non-nuisance grouping column in the output with its mostly meaningless summary stats; but I think it make sense to leave these out by default (which is the current behaviour). 

The goal of this PR is just to fix the bug in `GroupBy.apply`, leaving the behaviour of `GroupBy.describe` (and any other function) unchanged.

smithto1 replied:
Following this update the `.describe` method is no longer affected by this PR. 

https://github.com/pandas-dev/pandas/pull/35314#issuecomment-664193224

TomAugspurger commented on pandas/core/groupby/groupby.py:
```diff
@@ -1623,11 +1625,17 @@ def ohlc(self) -> DataFrame:
+            with _group_selection_context(self):
+                if self._group_selection is not None:
+                    cols = result.columns.intersection(self._group_selection)
+                    result = result.reindex(columns=cols)
```
This will perhaps break for duplicates in the columns. Can you check?

```python
In [13]: df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=['A', 'B', 'B'])

In [14]: df.groupby('A', as_index=False).describe()
Out[14]:
      A                                       B
  count mean std  min  25%  50%  75%  max count mean std  min  25%  50%  75%  max count mean std  min  25%  50%  75%  max
0   1.0  1.0 NaN  1.0  1.0  1.0  1.0  1.0   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0   1.0  3.0 NaN  3.0  3.0  3.0  3.0  3.0
1   1.0  4.0 NaN  4.0  4.0  4.0  4.0  4.0   1.0  5.0 NaN  5.0  5.0  5.0  5.0  5.0   1.0  6.0 NaN  6.0  6.0  6.0  6.0  6.0
```

smithto1 replied:
That is a good point @TomAugspurger . 

I've updated the code to handle this and added a test for it: `test_function.py::test_describe_with_duplicate_output_column_names`. None of the existing tests caught this.

smithto1 replied:
Following this update the `.describe` method is no longer affected by this PR. 

https://github.com/pandas-dev/pandas/pull/35314#issuecomment-664193224

I have kept the new test `test_function.py::test_describe_with_duplicate_output_column_names` as it makes sense to have a test for duplicate column names just to identify any future breaking changes.

TomAugspurger commented on pandas/core/groupby/groupby.py:
```diff
@@ -1623,11 +1625,17 @@ def ohlc(self) -> DataFrame:
-            return result.unstack()
+        result = self.apply(lambda x: x.describe(**kwargs))
+        if self.axis == 1:
+            return result.T
```
Was `axis=1` not affected by the same issue?

smithto1 replied:
AFAICT, no. 

This issue only came up when grouping by a column of non-nuisance values. `DataFrame.groupby` with `axis=1` doesn't apply to the rows the same way `axis=0` applies to the columns. If `axis=1` then `by` needs to be a column level-name or a mapping. 

See examples below. These all produce the same output on master and on PR branch. 

```
import numpy as np
import pandas as pd

df = pd.DataFrame(
    {
        "a": [99, 99, 99, 88, 88, 88],
        "b": [1, 2, 3, 4, 5, 6],
        "c": [10, 20, 30, 40, 50, 60],
    }
)

# axis=0 can group on column of values
df.groupby('a', axis=0).describe()

# axis=1 cannot group on row of values. Raises KeyError
df_T = df.T
df_T.groupby('a', axis=1)

# axis=1 works when 'by' is an index level name
df_T = df.set_index('a').T
df_T.groupby('a', axis=1).describe()

# axis=1 works when 'by' is a mapping
df_T = df.T
df_T.groupby(np.array([0,0,0,1,1,1]), axis=1).describe()

```
So `.describe` with `axis=1` does not seem to be affected by the original bug or by the changes. 

There are existing tests for `GroupBy.describe` with 'axis=1` and none of them were affected by the change. 

https://github.com/pandas-dev/pandas/blob/master/pandas/tests/groupby/test_function.py#L941

smithto1 replied:
Following this update the `.describe` method is no longer affected by this PR. 

https://github.com/pandas-dev/pandas/pull/35314#issuecomment-664193224

rhshadrach commented:
It seems to me that the root issue is `_agg_general` is calling `self._set_group_selection()` rather than using the context manager `with _group_selection_context(self)`. 

Wrapping `_agg_general` in this context manager, only two groupby tests fail. In each of these tests on master, if you comment out the above computations (e.g. `g.sum()`), the result changes. So I think that it is the tests that have an issue.

smithto1 committed:
fixed describe to work with duplicate cols

smithto1 committed:
merge fixed

smithto1 committed:
update comment

smithto1 committed:
context manager in agg_general

smithto1 committed:
remove hashed out line

smithto1 committed:
limited context manager in _make_wrapper

smithto1 commented:
> It seems to me that the root issue is `_agg_general` is calling `self._set_group_selection()` rather than using the context manager `with _group_selection_context(self)`.
> 
> Wrapping `_agg_general` in this context manager, the two issues handled by this PR are fixed and only two groupby tests fail. In each of these tests on master, if you comment out the above computations (e.g. `g.sum()`), the result changes. So I think that it is the tests that have an issue.
> 
> Update: The two tests that would need to be fixed with this suggestion are already done in this PR. The change to the shift transform test would not be necessary.

Thanks @rhshadrach , I think your recommendation is better and have gone with that. 

As well as `_agg_general`, I had to make the same update in `_make_wrapper` and `nth` which were also calling `self._set_group_selection`. I've replaced all of these with `with _group_selection_context(self):`. 

This approach means a method that calls `.apply` can still use `with _group_selection_context(self):` to limit the columns. So I don't need to include any changes to the `.describe` method to preserve it's behaviour. It is unaffected. 

My new test and the updates to the affected existing tests are still included.

smithto1 commented:
I've implemented the change recommended by @rhshadrach here
https://github.com/pandas-dev/pandas/pull/35314#issuecomment-664193224 

Rather than calling `self._reset_group_selection` inside `.apply`, I've found the three places where we were calling `self._set_group_selection` and replaced those with `with _group_selection_context(self):`. 

So this PR no longer has any impact `.describe` which we were discussing above. `.describe` is unchanged from it current code & behaviour in master/1.0.5.

smithto1 committed:
removed unrelated test

smithto1 committed:
update comment

TomAugspurger commented:
I'm not going to have a chance to look through this closely today. Given that we're in an RC period and this isn't a regression fix, I think we should target this for 1.1.1.

simonjayhawkins commented:
> I'm not going to have a chance to look through this closely today. Given that we're in an RC period and this isn't a regression fix, I think we should target this for 1.1.1.

sgtm

simonjayhawkins commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -1132,6 +1132,7 @@ Groupby/resample/rolling
 - Bug in :meth:`core.groupby.DataFrameGroupBy.transform` when ``func='nunique'`` and columns are of type ``datetime64``, the result would also be of type ``datetime64`` instead of ``int64`` (:issue:`35109`)
 - Bug in :meth:`DataFrame.groupby` raising an ``AttributeError`` when selecting a column and aggregating with ``as_index=False`` (:issue:`35246`).
 - Bug in :meth:`DataFrameGroupBy.first` and :meth:`DataFrameGroupBy.last` that would raise an unnecessary ``ValueError`` when grouping on multiple ``Categoricals`` (:issue:`34951`)
+- Bug in :meth:`DataFrameGroupBy.apply` where a non-nuisance grouping column would be dropped from the output columns if another groupby method was called before ``.apply()`` (:issue:`34656`)
```
can you move to bug fix section of doc/source/whatsnew/v1.1.1.rst

smithto1 replied:
Done

simonjayhawkins commented on pandas/tests/groupby/test_apply.py:
```diff
@@ -1014,3 +1014,32 @@ def test_apply_with_timezones_aware():
+        index=pd.Index([88, 99], name="a"),
+    )
+
+    # Check output wehn no other methods are called before .apply()
```
typo

smithto1 replied:
Done

simonjayhawkins reviewed (commented):
generally lgtm.

smithto1 committed:
Merge remote-tracking branch 'upstream/master' into issue34656

smithto1 committed:
whatsnew on v1.1.1

smithto1 committed:
comment typo

smithto1 committed:
amend comment to restart tests

jreback commented on doc/source/whatsnew/v1.1.1.rst:
```diff
@@ -40,6 +40,10 @@ Bug fixes
 
 -
 
+**Groupby/resample/rolling**
```
move to 1.2, generally we won't backport regular bugs, only regressions.

jreback commented on pandas/core/groupby/groupby.py:
```diff
@@ -990,28 +989,30 @@ def _agg_general(
+                    #  be trimmed by implementing cython funcs for more dtypes
+                    pass
+                else:
+                    raise
```
wow, codecov pointed that this is not covered, can you see if you can get a test to lands here (otherwise remove it)

smithto1 replied:
I could not find a way to reach this line from any public function call.  AFAICS anything that would raise a `NotImplementedError` that comes under the `else:`, also trips a `DataError` which catches it before this line.

So I've gone with removing it.

smithto1 committed:
whatsnew to 1.2.0

smithto1 committed:
resolve merge

smithto1 committed:
remove line that can't be tiggered by test

smithto1 committed:
restart tests

smithto1 commented:
@jreback I think this one is ready to be merged.

jreback commented:
lgtm. can you merge master again and ping on green (as merged a few other groupby patches recently).

smithto1 committed:
Merge remote-tracking branch 'upstream/master' into issue34656

jreback commented:
thanks @smithto1 (failure is unrelated)
